In [1]:
import numpy as np
np.set_printoptions(precision=4,suppress=True)
# precision - to set the number of decimal places the float will be printed
# Supress - if set True a float number will be displayed in numerical form else scientific form

### Example mxn matrix where m < n

In [2]:
a = np.array([[1,2,3,4],[1,1,2,3],[0,1,1,0]])

In [4]:
print(a)

[[1 2 3 4]
 [1 1 2 3]
 [0 1 1 0]]


### Apply SVD

In [5]:
u, s, vh = np.linalg.svd(a, full_matrices=True)

In [6]:
print(u)

[[ 0.8109 -0.0934  0.5776]
 [ 0.57    0.3493 -0.7437]
 [ 0.1323 -0.9324 -0.3365]]


In [7]:
u.dot(u.T)

array([[ 1.,  0.,  0.],
       [ 0.,  1., -0.],
       [ 0., -0.,  1.]])

In [8]:
vh.dot(vh.T)

array([[ 1.,  0.,  0., -0.],
       [ 0.,  1.,  0., -0.],
       [ 0.,  0.,  1., -0.],
       [-0., -0., -0.,  1.]])

### Hence u and vh are orthogonal matrices

In [9]:
s

array([6.7509, 1.1734, 0.2186])

### The first singular value of the matric is 6.7509, second singular value is 1.1734 and third singular value is 0.2186

In [10]:
print(u.shape, s.shape, vh.shape)

(3, 3) (3,) (4, 4)


In [11]:
print(s)

[6.7509 1.1734 0.2186]


### Since s is given simply as a 3-element numpy array, we need to reshape it into a 3x4 array with the 3 elements along its diagonal. Below is one way to do it - not the only way to do it

In [12]:
sd = np.diag(s) # makes a 3x3 diagonal martix with x along the diagonal

In [13]:
print(sd)

[[6.7509 0.     0.    ]
 [0.     1.1734 0.    ]
 [0.     0.     0.2186]]


In [14]:
sd.shape

(3, 3)

In [15]:
b = np.zeros((3,4))

In [16]:
print(b)

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


In [17]:
b[:,:-1] = sd

In [18]:
sigma = b
print(sigma)

[[6.7509 0.     0.     0.    ]
 [0.     1.1734 0.     0.    ]
 [0.     0.     0.2186 0.    ]]


In [19]:
print(np.dot(np.dot(u,sigma),vh)) # Reconstruction 'a' as product of 'u', 'sigma', 'vh'

[[1. 2. 3. 4.]
 [1. 1. 2. 3.]
 [0. 1. 1. 0.]]


### Currently the rank of a is 3 lets make a 2 rank approximation of matrix a

In [20]:
s

array([6.7509, 1.1734, 0.2186])

In [21]:
s = np.array([6.7509, 1.1734, 0])

In [22]:
sd = np.diag(s)
print(sd)

[[6.7509 0.     0.    ]
 [0.     1.1734 0.    ]
 [0.     0.     0.    ]]


In [23]:
b = np.zeros((3,4))
print(b)

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


In [24]:
b[:,:-1] = sd
sigma = b
print(sigma)

[[6.7509 0.     0.     0.    ]
 [0.     1.1734 0.     0.    ]
 [0.     0.     0.     0.    ]]


In [25]:
print(a)
print(np.dot(np.dot(u,sigma),vh))

[[1 2 3 4]
 [1 1 2 3]
 [0 1 1 0]]
[[ 1.0959  1.9567  3.0526  3.9542]
 [ 0.8764  1.0558  1.9322  3.0589]
 [-0.0559  1.0252  0.9693  0.0267]]


### Another Example - an mxn matrix where m > n

In [26]:
a = np.array([[1,2,3,4],[1,1,2,3],[0,1,1,0],[0,2,2,0],[0,5,5,0]])
print(a.shape)

(5, 4)


In [27]:
u, s, vh = np.linalg.svd(a)

In [28]:
print("u is :- ")
print(u)
print("\n s is :- ")
print(s)
print("\n vh is :- ")
print(vh)

u is :- 
[[-0.5185 -0.5983  0.6108  0.     -0.    ]
 [-0.3389 -0.512  -0.7893 -0.      0.    ]
 [-0.1433  0.1125 -0.0114  0.9169  0.3549]
 [-0.2866  0.225  -0.0229  0.2719 -0.8904]
 [-0.7166  0.5626 -0.0572 -0.2921  0.2852]]

 s is :- 
[9.2296 4.4454 0.2312 0.    ]

 vh is :- 
[[-0.0929 -0.6149 -0.7079 -0.3349]
 [-0.2498  0.375   0.1252 -0.8839]
 [-0.7718  0.3846 -0.3872  0.3264]
 [ 0.5774  0.5774 -0.5774 -0.    ]]
